Upgrade to egui v0.25.0 and miniquad v0.4

caspark opened the pull request:
This includes:
* @not-fl3's (i.e. your) https://github.com/not-fl3/egui-miniquad/pull/63
    * It would be much easier to review this PR if that PR were merged first. I don't think anything much I've done here depends on that PR (so feel free to rebase and drop this commit), but I wanted all this to work against the latest miniquad version, so I included your PR because of that. Also, that PR has been open for 6 months now, so I figured why not include it :)
* @emilk's https://github.com/not-fl3/egui-miniquad/pull/64
    * (well, not the exact commit from that PR: I copied over the 2 lines of changes from it, bumped to latest miniquad 4.x alpha, and fixed the 2 broken examples in this crate)
* and 2 more commits to update to egui `0.24.1` and `0.25.0`.

I've tested a little bit and "things seem to work" - at least the obvious stuff like rendering, mouse and keyboard input handling, etc.

Things I'm not sure about:
* The flickering between the color-test gradients that @emilk noted in https://github.com/not-fl3/egui-miniquad/pull/64 is still there. It only appears between two gradients.. is that a big deal? Not for me, but it might be for others.
* ~High DPI screens support. [Egui 0.24.0 revamped](https://github.com/emilk/egui/blob/master/CHANGELOG.md#0240---2023-11-23---multi-viewport) how pixels per point is calculated and used. I made a best effort attempt to integrate this correctly but I couldn't find any big picture docs and I don't use a high DPI screen so I'm not sure it's done correctly.~
    * ~And it looks like egui can suggest DPI changes back to the egui backend now as well, based on the user zooming in or something like that? I didn't implement this side of things at all because I don't need it.~
    * Update: this should now be implemented properly thanks to some helpful pointers from @emilk .

So.. it's not a rock solid PR, but I'm raising it anyway to save others doing the same work if they don't care about these limitations.

emilk commented on src/lib.rs:
```diff
@@ -184,9 +184,10 @@ impl EguiMq {
-            repaint_after: _, // miniquad always runs at full framerate
             textures_delta,
             shapes,
+            pixels_per_point, //TODO not handling change in pixels per point (check zooming in/out behavior in egui)
```
This is egui telling you what its (current) definition of a "point" is. It is the value that should be passed to `tessellate`

emilk commented on src/lib.rs:
```diff
@@ -210,25 +211,25 @@ impl EguiMq {
+    pub fn draw(&mut self, mq_ctx: &mut dyn mq::RenderingBackend) {
         if let Some(shapes) = self.shapes.take() {
-            let meshes = self.egui_ctx.tessellate(shapes);
+            let meshes = self.egui_ctx.tessellate(shapes, self.native_dpi_scale);
```
Here we should use the `pixels_per_point` reported by egui, because that can differ from `native_dpi_scale` if the user has bit Ctrl+/-

emilk reviewed (approved):
Thank you! Looks good, except for the pixels-per-point thing

not-fl3 commented:
Give me about a week, I am about to publish miniquad-0.4 without alpha. 

_(yes, I was about to publish it since forever, but now I really am!)_

caspark commented:
* Thanks for the pointers @emilk - got the points per pixel stuff fixed now. (I suspected it was meant to be that simple, but when I tried it nothing happened, which was because of the next bullet)
* @not-fl3 I've also added support for punctuation keys that were missing, most notably plus and equals!
* I've also replaced the manual override of pixels per point in the `demo` example with manually setting zoom factor instead; otherwise changing zoom factor in egui with its built in keyboard bindings will immediately be reset to 1.0 zoom next frame.

With all 3 of those things fixed, you can now zoom the `demo` example in and out (and reset zoom) as expected with Ctrl/Command +/-/0.

So this PR is now B+ quality, with the only remaining thing being the gap between the gradients (which, again, I can live with).

emilk commented on examples/demo.rs:
```diff
@@ -5,37 +5,45 @@ struct Stage {
             egui_demo_windows: Default::default(),
             color_test: Default::default(),
-            pixels_per_point: ctx.dpi_scale(),
+            zoom_factor: mq::window::dpi_scale(),
```
This is wrong - the default zoom factor in egui is always 1.0, but the default dpi-scale could be e.g. 2.0 for a retina screen.

But I suggest we just remove the `zoom_factor` stuff from the demo, since this is now built-in to egui with `ctrl+/0`.

caspark replied:
Whoops, fixed.

I'll let @not-fl3 decide whether to remove the zoom factor stuff - I'm just preserving the existing functionality here.

emilk commented:
Btw, a new egui is coming next week too :)

not-fl3 commented:
> Give me about a week, I am about to publish miniquad-0.4 without alpha.
> 
> _(yes, I was about to publish it since forever, but now I really am!)_

> commented 3 weeks ago

anyway, miniquad is released, no more github/crates differences nor -alpha versions! 
Most changes are internal for metal backend, so this PR should(mostly) just work.

caspark commented:
Yay!

I've updated this PR to target miniquad 0.4.0 now, so: time to hit merge?

(I know there's egui 0.26 out already, but maybe let's just get this merged first)
